### PR TITLE
pageserver: simplify LayerAccessStats

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

LayerAccessStats contains a lot of detail that we don't use: short histories of most recent accesses, specifics on what kind of task accessed a layer, etc.  This is all stored inside a Mutex, which is locked every time something accesses a layer.

## Summary of changes

- Store timestamps at a very low resolution (to the nearest second), sufficient for use on the timescales of eviction.
- Pack access time and last residence change time into a single u64
- Use the high bits of the u64 for other flags, including the new layer visibility concept.
- Simplify the external-facing model for access stats to just include what we now track.

Note that the `HistoryBufferWithDropCounter` is removed here because it is no longer used.  I do not dislike this type, we just happen not to use it for anything else at present.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by jcsp** [comment]
The test runner has decided that test_partial_eviction is flaky (probably because of this PR), so check for that when interpreting results.

Example failure: https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8431/10040337265/index.html#suites/0e58fb04d9998963e98e45fe1880af7d/79096da9c207613e/retries

**Review by yliang412 — APPROVED** [review]
Overall looks good. For `test_partial_eviction`, it does seem the ratio is close to 0.06? I wonder if it is close enough?

File: pageserver/src/tenant/storage_layer.rs [hunk 1]
```diff
@@ -458,40 +447,32 @@ pub enum ValueReconstructResult {
...
-#[derive(Debug, Clone, Copy)]
-pub(crate) struct LayerAccessStatFullDetails {
-    pub(crate) when: SystemTime,
-    pub(crate) task_kind: TaskKind,
-    pub(crate) access_kind: LayerAccessKind,
-}
+pub(crate) struct LayerAccessStats(std::sync::atomic::AtomicU64);
 
 #[derive(Clone, Copy, strum_macros::EnumString)]
 pub enum LayerAccessStatsReset {
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant/storage_layer.rs]: Good spot, removed in d354d9b18306574f0004c7b93dd1a5c687d1b55e

File: test_runner/regress/test_threshold_based_eviction.py [hunk 2]
```diff
@@ -52,8 +52,8 @@ def test_threshold_based_eviction(
         "kind": "NoEviction"
     }
 
-    eviction_threshold = 5
-    eviction_period = 1
+    eviction_threshold = 10
+    eviction_period = 2
```
Comment (problame) [inline test_runner/regress/test_threshold_based_eviction.py:56]: Should this use the `ATIME_RESOLUTION` variable introduce in `test_disk_usage_based_eviction` or it is by accident that this is also `2`?

**Review by problame — APPROVED** [review]
We can remove the `heapless` crate as a dependency, I think.

---

Why do we need `::Uninitalized`? This compiles:

```patch
commit 060a4fbc499829fe57f07baee72d41be80db464d
Author: Christian Schwarz <christian@neon.tech>
Date:   Tue Jul 23 18:07:48 2024 +0200

    remove unused LayerVisibilityHint::Uninitialized

diff --git a/pageserver/src/tenant/storage_layer.rs b/pageserver/src/tenant/storage_layer.rs
index efee4477a..0518f904b 100644
--- a/pageserver/src/tenant/storage_layer.rs
+++ b/pageserver/src/tenant/storage_layer.rs
@@ -451,7 +451,7 @@ pub enum ValueReconstructResult {
 /// than an authoritative value, so that we do not have to update it synchronously when changing the visibility
 /// of layers (for example when creating a branch that makes some previously covered layers visible).  It should
 /// be used for cache management but not for correctness-critical checks.
-#[derive(Default, Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub(crate) enum LayerVisibilityHint {
     /// A Visible layer might be read while serving a read, because there is not an image layer between it
     /// and a readable LSN (the tip of the branch or a child's branch point)
@@ -460,11 +460,6 @@ pub(crate) enum LayerVisibilityHint {
     /// a branch or ephemeral endpoint at an LSN below the layer that covers this.
     #[allow(unused)]
     Covered,
-    /// Calculating layer visibilty requires I/O, so until this has happened layers are loaded
-    /// in this state.  Note that newly written layers may be called Visible immediately, this uninitialized
-    /// state is for when existing layers are constructed while loading a timeline.
-    #[default]
-    Uninitialized,
 }
 
 pub(crate) struct LayerAccessStats(std::sync::atomic::AtomicU64);
@@ -629,7 +624,7 @@ impl LayerAccessStats {
     pub(crate) fn set_visibility(&self, visibility: LayerVisibilityHint) {
         let value = match visibility {
             LayerVisibilityHint::Visible => 0x1 << Self::VISIBILITY_SHIFT,
-            LayerVisibilityHint::Covered | LayerVisibilityHint::Uninitialized => 0x0,
+            LayerVisibilityHint::Covered => 0x0,
         };
 
         self.write_bits(0x1 << Self::VISIBILITY_SHIFT, value);
```

File: test_runner/regress/test_threshold_based_eviction.py (hunk 2 above)
Comment (jcsp) [reply test_runner/regress/test_threshold_based_eviction.py:56]: It's a coincidence: these are just arbitrarily 2x the old values

**Comment by jcsp** [comment]
>We can remove the heapless crate as a dependency, I think.

yay.  Done in f22109dc4

> Why do we need ::Uninitalized? This compiles:

The type comes from a branch where it's meaningful... I'll figure out if I still need it and remove if not in the next PR for layer visibility

**Pull request merged**